src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Types/TypeArray.cs

```diff
@@ -79,6 +79,7 @@ public override bool Equals(object obj)
             public override int GetHashCode() => _hashCode;
         }
 
+	// The RuntimeBinder uses a global lock when Binding that keeps this dictionary safe
```

Comment:
@VSadov, can/should we `Debug.Assert(Monitor.IsEntered(lockObjectHere))` that the appropriate lock is held while accessing the dictionary?

Reply:
We could do that by adding a method on the RuntimeBinder struct that calls
`Debug.Assert(System.Threading.Monitor.IsEntered(s_bindLock))` and calling that method from all the places where the dictionaries are accessed.

We could even make the method conditionally compiled on DEBUG... so that we don't mess with the release build.

Reply:
Done.